design/external-processing-design.md

```diff
@@ -0,0 +1,660 @@
+	//
+	// +optional
+	// +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms)|(\d*(\.\d*)?us)|(\d*(\.\d*)?µs)|(\d*(\.\d*)?ns))+|infinity|infinite)$`
+	ResponseTimeout string `json:"responseTimeout,omitempty"`
+
+	// If FailOpen is true, the client request is forwarded to the upstream service
+	// even if the server fails to respond. This field should not be
+	// set in most cases.
+	//
+	// +optional
+	FailOpen bool `json:"failOpen,omitempty"`
+}
+
+
+// ExtProc defines the envoy External Processing filter which allows an external service to act on HTTP traffic in a flexible way
+// The external server must implement the v3 Envoy external processing GRPC protocol
+// (https://www.envoyproxy.io/docs/envoy/v1.27.0/api-v3/extensions/filters/http/ext_proc/v3/ext_proc.proto).
+type ExtProc struct {
+	// GRPCService configure the gRPC service that the filter will communicate with.
+	//
+	// +optional
+	GRPCService *GRPCService `json:"grpcService,omitempty"`
+
+	// ProcessingMode describes which parts of an HTTP request and response are sent to a remote server
+	// and how they are delivered.
+	//
+	// +optional
+	ProcessingMode *ProcessingMode `json:"processingMode,omitempty"`
+
+	// MutationRules specifies what headers may be manipulated by a processing filter.
+	// This set of rules makes it possible to control which modifications a filter may make.
+	//
+	// for Overrides is must be nil
+	//
+	// +optional
+	MutationRules *HeaderMutationRules `json:"mutationRules,omitempty"`
+
+	// If true, the filter config processingMode can be overridden by the response message from the external processing server `mode_override``.
+	// If false, `mode_override` API in the response message will be ignored.
+	//
+	// +optional
+	AllowModeOverride bool `json:"allowModeOverride,omitempty"`
+}
+
+// ExternalProcessor defines a external processing filter and the policy for fine-grained at VirutalHost and/or Route level.
+type ExternalProcessor struct {
+	// Processor defines a external processing filter which allows an external service to act on HTTP traffic in a flexible way.
+	//
+	// +optional
+	Processor *ExtProc `json:"processor,omitempty"`
+
+	// When true, this field disables the external processor: (neither global nor virtualHost)
+	// for the scope of the policy.
+	//
+	// if both Disabled and Processor are set. use disabled.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+}
+
+// ExtProcPolicy modifies how requests/responses are operated.
+type ExtProcPolicy struct {
+	// When true, this field disables the specific client request external processor
+	// for the scope of the policy.
+	//
+	// if both disabled and overrides are set. use disabled.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Overrides aspects of the configuration for this route.
+	//
+	// +optional
+	Overrides *ExtProc `json:"overrides,omitempty"`
+}
```

Comment:
Looking at the implementation in PR I feel the HTTPProxy struct extproc options feel a bit fragmented 

is there a reason we cant bundle them up into one struct member on the `type VirtualHost struct {}`

This is what i was thinking 

```
type ExternalProcessingServer struct {
	// ExtensionServiceRef specifies the extension resource that will handle the client requests.
	//
	// +optional
	ExtensionServiceRef ExtensionServiceReference `json:"extensionRef,omitempty"`

	// ResponseTimeout sets how long the proxy should wait for responses.
	// Timeout durations are expressed in the Go [Duration format](https://godoc.org/time#ParseDuration).
	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
	// The string "infinity" is also a valid input and specifies no timeout.
	//
	// +optional
	// +kubebuilder:validation:Pattern=`^(((\d*(\.\d*)?h)|(\d*(\.\d*)?m)|(\d*(\.\d*)?s)|(\d*(\.\d*)?ms)|(\d*(\.\d*)?us)|(\d*(\.\d*)?µs)|(\d*(\.\d*)?ns))+|infinity|infinite)$`
	ResponseTimeout string `json:"responseTimeout,omitempty"`

	// If FailOpen is true, the client request is forwarded to the upstream service
	// even if the server fails to respond. This field should not be
	// set in most cases.
	//
	// +optional
	FailOpen bool `json:"failOpen,omitempty"`

	// ProcessingMode describes which parts of an HTTP request and response are sent to a remote server
	// and how they are delivered.
	//
	// +optional
	ProcessingMode *ProcessingMode `json:"processingMode,omitempty"`

	// MutationRules specifies what headers may be manipulated by a processing filter.
	// This set of rules makes it possible to control which modifications a filter may make.
	//
	// for Overrides is must be nil
	//
	// +optional
	MutationRules *HeaderMutationRules `json:"mutationRules,omitempty"`

	// If true, the filter config processingMode can be overridden by the response message from the external processing server `mode_override``.
	// If false, `mode_override` API in the response message will be ignored.
	//
	// +optional
	AllowModeOverride bool `json:"allowModeOverride,omitempty"`
	
	ExtProcPolicy *ExtProcPolicy  `json:"extProcPolicy,omitempty"`
}


...


type VirtualHost struct {
...
ExtProcServer ExternalProcessingServer `json:"extProcServer,omitempty"`
...
}
```

Reply:
this also keeps the UX similar to what we have for ratelimiting and extauth

Reply:
@clayton-gonsalves PTAL, BTW let our review the design first, then the implemention ;-)